File: iosApp/iosApp/ui/swap/PreviewableSwapTruthSource.swift
//
//  PreviewableSwapTruthSource.swift
//  iosApp
//
//  Created by jimmyt on 8/15/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import web3swift

/**
 A `UISwapTruthSource` implementation used for previewing user interfaces.
 */
class PreviewableSwapTruthSource: UISwapTruthSource {
    /**
     Initializes a new `PreviewableSwapTruthSource` with sample swaps.
     */
    init() {
        swaps = Swap.sampleSwaps
    }
    
    /**
     A dictionary mapping swap IDs (as `UUID`s) to `Swap`s, which is the single source of truth for all swap-related data.
     */
    @Published var swaps: [UUID : Swap]
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func fillSwap(swap: Swap) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func createApproveTokenTransferToFillSwapTransaction(
        swap: Swap,
        createdTransactionHandler: @escaping (EthereumTransaction) -> Void,
        errorHandler: @escaping (Error) -> Void
    ) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func approveTokenTransferToFillSwap(
        swap: Swap,
        approveTokenTransferToFillSwapTransaction: EthereumTransaction?
    ) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func createFillSwapTransaction(
        swap: Swap,
        createdTransactionHandler: @escaping (EthereumTransaction) -> Void,
        errorHandler: @escaping (Error) -> Void
    ) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func fillSwap(
        swap: Swap,
        swapFillingTransaction: EthereumTransaction?
    ) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func reportPaymentSent(swap: Swap) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func createReportPaymentSentTransaction(swap: Swap, createdTransactionHandler: @escaping (EthereumTransaction) -> Void, errorHandler: @escaping (Error) -> Void ) {}

    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func reportPaymentSent(swap: Swap, reportPaymentSentTransaction: EthereumTransaction?) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func reportPaymentReceived(swap: Swap) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func createReportPaymentReceivedTransaction(swap: Swap, createdTransactionHandler: @escaping (EthereumTransaction) -> Void, errorHandler: @escaping (Error) -> Void ) {}

    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func reportPaymentReceived(swap: Swap, reportPaymentReceivedTransaction: EthereumTransaction?) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func closeSwap(swap: Swap) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func createCloseSwapTransaction(swap: Swap, createdTransactionHandler: @escaping (EthereumTransaction) -> Void, errorHandler: @escaping (Error) -> Void ) {}

    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UISwapTruthSource`.
     */
    func closeSwap(swap: Swap, closeSwapTransaction: EthereumTransaction?) {}
    
}

File: androidApp/src/main/java/com/commuto/interfacemobile/android/ui/swap/PreviewableSwapTruthSource.kt
package com.commuto.interfacemobile.android.ui.swap

import androidx.compose.runtime.snapshots.SnapshotStateMap
import com.commuto.interfacemobile.android.offer.Offer
import com.commuto.interfacemobile.android.swap.Swap
import com.commuto.interfacemobile.android.ui.offer.UIOfferTruthSource
import org.web3j.crypto.RawTransaction
import java.util.*

/**
 * A [UISwapTruthSource] implementation used for previewing user interfaces.
 *
 * @property swaps A [SnapshotStateMap] mapping  [UUID]s to [Swap]s, which acts as a single source of truth for all
 * swap-related data.
 */
class PreviewableSwapTruthSource : UISwapTruthSource {
    override var swaps = SnapshotStateMap<UUID, Swap>().also { map ->
        Swap.sampleSwaps.map {
            map[it.id] = it
        }
    }

    /**
     * Does nothing, since this class is only used for previewing user interfaces.
     */
    override fun addSwap(swap: Swap) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UISwapTruthSource].
     */
    override fun fillSwap(swap: Swap) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UISwapTruthSource].
     */
    override fun reportPaymentSent(swap: Swap) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UISwapTruthSource].
     */
    override fun createReportPaymentSentTransaction(
        swap: Swap,
        createdTransactionHandler: (RawTransaction) -> Unit,
        exceptionHandler: (Exception) -> Unit
    ) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UISwapTruthSource].
     */
    override fun reportPaymentSent(swap: Swap, reportPaymentSentTransaction: RawTransaction?) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UISwapTruthSource].
     */
    override fun reportPaymentReceived(swap: Swap) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UISwapTruthSource].
     */
    override fun createReportPaymentReceivedTransaction(
        swap: Swap,
        createdTransactionHandler: (RawTransaction) -> Unit,
        exceptionHandler: (Exception) -> Unit
    ) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UISwapTruthSource].
     */
    override fun reportPaymentReceived(swap: Swap, reportPaymentReceivedTransaction: RawTransaction?) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UISwapTruthSource].
     */
    override fun closeSwap(swap: Swap) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UISwapTruthSource].
     */
    override fun createCloseSwapTransaction(
        swap: Swap,
        createdTransactionHandler: (RawTransaction) -> Unit,
        exceptionHandler: (Exception) -> Unit
    ) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UISwapTruthSource].
     */
    override fun closeSwap(swap: Swap, closeSwapTransaction: RawTransaction?) {}
}